File: apps/web/src/lib/organizations/settlement-wallet.ts
import { and, asc, eq } from "drizzle-orm";
import { db } from "@/lib/db";
import {
  organizationMembers,
  organizations,
  organizationSettlementWallets,
  type Organization,
  type OrganizationSettlementWallet,
} from "@/lib/db/schema";

export async function getSettlementWallet(
  organizationId: string,
  environment: Organization["environment"],
) {
  const [wallet] = await db
    .select()
    .from(organizationSettlementWallets)
    .where(
      and(
        eq(organizationSettlementWallets.organizationId, organizationId),
        eq(organizationSettlementWallets.environment, environment),
      ),
    )
    .limit(1);

  return wallet ?? null;
}

export function settlementWalletNotConfiguredMessage(
  environment: Organization["environment"],
) {
  if (environment === "production") {
    return "Production settlement wallet is not configured. Open Settings → Settlement Wallet and connect your Mainnet wallet.";
  }

  return "Sandbox settlement wallet is not configured. Open Settings → Settlement Wallet and connect your Testnet wallet.";
}

export async function requireSettlementWallet(
  organizationId: string,
  environment: Organization["environment"],
) {
  const wallet = await getSettlementWallet(organizationId, environment);

  if (!wallet) {
    throw new Error(settlementWalletNotConfiguredMessage(environment));
  }

  return wallet;
}

export async function organizationHasSettlementWallet(
  organizationId: string,
  environment: Organization["environment"],
) {
  const wallet = await getSettlementWallet(organizationId, environment);
  return Boolean(wallet);
}

export async function upsertSettlementWallet(input: {
  organizationId: string;
  environment: Organization["environment"];
  stellarAddress: string;
  walletProvider?: string | null;
}): Promise<OrganizationSettlementWallet> {
  const existing = await getSettlementWallet(
    input.organizationId,
    input.environment,
  );

  if (existing) {
    const [wallet] = await db
      .update(organizationSettlementWallets)
      .set({
        stellarAddress: input.stellarAddress,
        walletProvider: input.walletProvider ?? null,
        connectedAt: new Date(),
        updatedAt: new Date(),
      })
      .where(eq(organizationSettlementWallets.id, existing.id))
      .returning();

    return wallet;
  }

  const [wallet] = await db
    .insert(organizationSettlementWallets)
    .values({
      organizationId: input.organizationId,
      environment: input.environment,
      stellarAddress: input.stellarAddress,
      walletProvider: input.walletProvider ?? null,
    })
    .returning();

  return wallet;
}

export async function getPrimaryOrganizationForUser(userId: string) {
  const [membership] = await db
    .select()
    .from(organizationMembers)
    .where(eq(organizationMembers.userId, userId))
    .orderBy(asc(organizationMembers.createdAt))
    .limit(1);

  if (!membership) {
    return null;
  }

  const [organization] = await db
    .select()
    .from(organizations)
    .where(eq(organizations.id, membership.organizationId))
    .limit(1);

  return organization ?? null;
}

export async function userHasSettlementWallet(userId: string) {
  const organization = await getPrimaryOrganizationForUser(userId);

  if (!organization) {
    return false;
  }

  return organizationHasSettlementWallet(
    organization.id,
    organization.environment,
  );
}

export async function getOrganizationForMember(
  organizationId: string,
  userId: string,
) {
  const [membership] = await db
    .select()
    .from(organizationMembers)
    .where(
      and(
        eq(organizationMembers.organizationId, organizationId),
        eq(organizationMembers.userId, userId),
      ),
    )
    .limit(1);

  if (!membership) {
    return null;
  }

  const [organization] = await db
    .select()
    .from(organizations)
    .where(eq(organizations.id, membership.organizationId))
    .limit(1);

  return organization ?? null;
}

File: apps/api/internal/service/organizations/settlement_wallet.go
package organizations

import (
	"context"
	"errors"
	"time"

	"github.com/jackc/pgx/v5"
)

// SettlementWallet ports organization_receiving_wallets rows used by settlement/receiving wallet APIs.
type SettlementWallet struct {
	ID             string    `json:"id"`
	OrganizationID string    `json:"organizationId"`
	Environment    string    `json:"environment"`
	StellarAddress string    `json:"stellarAddress"`
	WalletProvider *string   `json:"walletProvider"`
	ConnectedAt    time.Time `json:"connectedAt"`
	UpdatedAt      time.Time `json:"updatedAt"`
}

func (s *Service) GetSettlementWallet(ctx context.Context, organizationID, environment string) (*SettlementWallet, error) {
	var w SettlementWallet
	err := s.pool.QueryRow(ctx, `
		SELECT id, organization_id, environment, stellar_address, wallet_provider, connected_at, updated_at
		FROM organization_receiving_wallets
		WHERE organization_id = $1 AND environment = $2
		LIMIT 1`, organizationID, environment).Scan(
		&w.ID, &w.OrganizationID, &w.Environment, &w.StellarAddress, &w.WalletProvider, &w.ConnectedAt, &w.UpdatedAt,
	)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &w, nil
}

func (s *Service) UpsertSettlementWallet(ctx context.Context, organizationID, environment, stellarAddress string, walletProvider *string) (*SettlementWallet, error) {
	existing, err := s.GetSettlementWallet(ctx, organizationID, environment)
	if err != nil {
		return nil, err
	}
	if existing != nil {
		var w SettlementWallet
		err := s.pool.QueryRow(ctx, `
			UPDATE organization_receiving_wallets
			SET stellar_address = $1, wallet_provider = $2, connected_at = NOW(), updated_at = NOW()
			WHERE id = $3
			RETURNING id, organization_id, environment, stellar_address, wallet_provider, connected_at, updated_at`,
			stellarAddress, walletProvider, existing.ID,
		).Scan(&w.ID, &w.OrganizationID, &w.Environment, &w.StellarAddress, &w.WalletProvider, &w.ConnectedAt, &w.UpdatedAt)
		if err != nil {
			return nil, err
		}
		return &w, nil
	}

	var w SettlementWallet
	err = s.pool.QueryRow(ctx, `
		INSERT INTO organization_receiving_wallets (organization_id, environment, stellar_address, wallet_provider)
		VALUES ($1, $2, $3, $4)
		RETURNING id, organization_id, environment, stellar_address, wallet_provider, connected_at, updated_at`,
		organizationID, environment, stellarAddress, walletProvider,
	).Scan(&w.ID, &w.OrganizationID, &w.Environment, &w.StellarAddress, &w.WalletProvider, &w.ConnectedAt, &w.UpdatedAt)
	if err != nil {
		return nil, err
	}
	return &w, nil
}

func SerializeSettlementWallet(w *SettlementWallet) map[string]any {
	if w == nil {
		return nil
	}
	return map[string]any{
		"stellarAddress": w.StellarAddress,
		"environment":    w.Environment,
		"walletProvider": w.WalletProvider,
		"connectedAt":    w.ConnectedAt,
	}
}
